src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -130,6 +138,117 @@ worker_fix_pre_citus10_partitioned_table_constraint_names(PG_FUNCTION_ARGS)
+ * indexes on each partition shard of the parent shards. If index name is too long, it
+ * truncates the name and adds _idx postfix to it. However, when truncating the name, the
+ * shardId of the partition shard can be lost. This may result in the same index name used for
+ * the partition shell table and one of the partition shards.
+ * For more details, check issue #4962 https://github.com/citusdata/citus/issues/4962
+ *
+ * fix_partition_shard_index_names renames indexes of shards of partition tables to include
+ * the shardId at the end of the name, regardless of whether index name was long or short
+ * As a result there will be no index name ending in _idx, rather all will end in _{shardid}
+ */
+Datum
+fix_partition_shard_index_names(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	EnsureCoordinator();
+
+	Relation relation = try_relation_open(relationId, AccessShareLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("could not fix partition shard index names: "
+							   "relation does not exist")));
+	}
+
+	if (relation->rd_rel->relkind != RELKIND_PARTITIONED_TABLE)
+	{
+		relation_close(relation, AccessShareLock);
+		ereport(ERROR, (errmsg("could not fix partition shard index names: "
+							   "relation is not partitioned")));
+	}
+
+	if (!IsCitusTable(relationId))
+	{
+		relation_close(relation, AccessShareLock);
+		ereport(ERROR, (errmsg("fix_partition_shard_index_names can "
+							   "only be called for distributed partitioned tables")));
+	}
+
+	List *taskList = CreateFixPartitionShardIndexNamesTaskList(relationId);
+
+	/* do not do anything if there are no index names to fix */
+	if (taskList != NIL)
+	{
+		bool localExecutionSupported = true;
+		ExecuteUtilityTaskList(taskList, localExecutionSupported);
+	}
+
+	relation_close(relation, AccessShareLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * worker_fix_partition_shard_index_names fixes the index names of child indexes
+ * given a parent index.
+ * The parent index should be the index of a shard of a distributed partitioned table.
+ * The child indexes should be the indexes of shards of the partitions of the distributed
+ * partitioned table from above.
+ */
+Datum
+worker_fix_partition_shard_index_names(PG_FUNCTION_ARGS)
+{
+	Oid shardIndexId = PG_GETARG_OID(0);
+
+	if (!has_subclass(shardIndexId))
+	{
+		ereport(ERROR, (errmsg("could not fix child index names: "
+							   "index is not partitioned")));
+	}
+
+	List *partitionShardIndexIds = find_inheritance_children(shardIndexId,
+															 ShareRowExclusiveLock);
+	Oid partitionShardIndexId = InvalidOid;
+	foreach_oid(partitionShardIndexId, partitionShardIndexIds)
+	{
+		/* check if this index name ends with shardid, if not rename it */
+		char *partitionShardIndexName = get_rel_name(partitionShardIndexId);
+		bool missing_ok = true;
+		if (ExtractShardIdFromTableName(partitionShardIndexName, missing_ok) ==
+			INVALID_SHARD_ID)
+		{
```

Comment:
I started to realize that liberally changing the index name on the shards (e.g., not doing for the coordinator shell table) would break the further operations on indexes. Can we/should we synchronize the index names on the coordinator?


```SQL

CREATE TABLE parent_table_with_very_long_name (
    dist_col integer,
    long_named_integer_col integer,
    long_named_part_col timestamp
) PARTITION BY RANGE (long_named_part_col);

CREATE TABLE child_table_with_long_long_long_long_name
PARTITION OF parent_table_with_very_long_name
FOR VALUES FROM ('2010-01-01') TO ('2015-01-01');

SELECT create_distributed_table('parent_table_with_very_long_name', 'dist_col');

CREATE INDEX ON parent_table_with_very_long_name
USING btree (long_named_integer_col, long_named_part_col);


-- fails as expected
select start_metadata_sync_to_node('localhost', 9700);   
ERROR:  relation "child_table_with_long_long_lo_long_named_integer_col_long_n_idx" already exists


SELECT fix_partition_shard_index_names('parent_table_with_very_long_name'::regclass);

-- now we lost the connection between the index on the distriibuted table and shards
alter index child_table_with_long_long_lo_long_named_integer_col_long_n_idx rename to smth_smaller;
ERROR:  relation "public.child_table_with_long_long_lo_long_named_intege_9784e088_102810" does not exist


select start_metadata_sync_to_node('localhost', 9700);   

-- same even after metadata is synced
alter index child_table_with_long_long_lo_long_named_integer_col_long_n_idx rename to smth_smaller;
ERROR:  relation "public.child_table_with_long_long_lo_long_named_intege_9784e088_102871" does not exist
```

Reply:
Thanks, working on this